File: frontend/lib/features/tenant_facility/domain/entities/department_similarity.dart
import 'package:hosspi_hms/features/tenant_facility/domain/entities/tenant_facility_setup.dart';
import 'package:hosspi_hms/features/tenant_facility/domain/entities/tenant_similarity.dart';

const int departmentSimilarityThreshold = tenantSimilarityThreshold;
const int departmentNameWeight = 50;
const int departmentShortNameWeight = 20;
const int departmentTypeWeight = 20;
const int departmentStatusWeight = 10;

typedef DepartmentFieldComparisonStatus = TenantFieldComparisonStatus;

final class DepartmentFieldComparison {
  const DepartmentFieldComparison({
    required this.field,
    this.inputValue,
    this.candidateValue,
    this.score,
    required this.status,
  });

  final String field;
  final String? inputValue;
  final String? candidateValue;
  final int? score;
  final DepartmentFieldComparisonStatus status;
}

final class DepartmentSimilarityProposedValues {
  const DepartmentSimilarityProposedValues({
    required this.name,
    required this.shortName,
    required this.type,
    required this.isActive,
  });

  final String name;
  final String shortName;
  final DepartmentSetupType type;
  final bool isActive;
}

final class DepartmentSimilarityMatch {
  const DepartmentSimilarityMatch({
    required this.department,
    required this.score,
    required this.reasons,
    required this.isExact,
    this.exactNameConflict = false,
    this.nameScore,
    this.shortNameScore,
    this.typeScore,
    this.statusScore,
    this.fieldComparisons = const <DepartmentFieldComparison>[],
  });

  final DepartmentProfile department;
  final int score;
  final List<String> reasons;
  final bool isExact;
  final bool exactNameConflict;
  final int? nameScore;
  final int? shortNameScore;
  final int? typeScore;
  final int? statusScore;
  final List<DepartmentFieldComparison> fieldComparisons;
}

final class DepartmentDuplicateCheckResult {
  const DepartmentDuplicateCheckResult({
    this.exactNameConflict = false,
    this.similarMatches = const <DepartmentSimilarityMatch>[],
  });

  final bool exactNameConflict;
  final List<DepartmentSimilarityMatch> similarMatches;

  bool get hasExactConflict => exactNameConflict;

  List<DepartmentSimilarityMatch> get nonExactSimilarMatches {
    return similarMatches
        .where((DepartmentSimilarityMatch match) => !match.isExact)
        .toList(growable: false);
  }

  List<DepartmentSimilarityMatch> get overridableMatches {
    return similarMatches
        .where((DepartmentSimilarityMatch match) => !match.exactNameConflict)
        .toList(growable: false);
  }
}

String normalizeDepartmentName(String value) => normalizeTenantName(value);

String resolveDepartmentShortName(String name, String? shortName) {
  final String trimmedShort = (shortName ?? '').trim();
  if (trimmedShort.isNotEmpty) {
    return trimmedShort;
  }
  return name.trim();
}

int compositeDepartmentSimilarityScore({
  int? nameScore,
  int? shortNameScore,
  int? typeScore,
  int? statusScore,
}) {
  var weightedTotal = 0;
  var weightSum = 0;

  void include(int? score, int weight) {
    if (score == null) {
      return;
    }
    weightedTotal += score * weight;
    weightSum += weight;
  }

  include(nameScore, departmentNameWeight);
  include(shortNameScore, departmentShortNameWeight);
  include(typeScore, departmentTypeWeight);
  include(statusScore, departmentStatusWeight);

  if (weightSum == 0) {
    return 0;
  }
  return (weightedTotal / weightSum).round();
}

DepartmentFieldComparisonStatus _comparisonStatus(
  int? score, {
  bool exact = false,
}) {
  if (exact || score == 100) {
    return DepartmentFieldComparisonStatus.match;
  }
  if (score == null) {
    return DepartmentFieldComparisonStatus.missing;
  }
  if (score >= departmentSimilarityThreshold) {
    return DepartmentFieldComparisonStatus.similar;
  }
  return DepartmentFieldComparisonStatus.different;
}

DepartmentFieldComparison _fieldComparison({
  required String field,
  String? inputValue,
  String? candidateValue,
  int? score,
  bool exact = false,
}) {
  return DepartmentFieldComparison(
    field: field,
    inputValue: inputValue?.trim().isEmpty == true ? null : inputValue?.trim(),
    candidateValue: candidateValue?.trim().isEmpty == true
        ? null
        : candidateValue?.trim(),
    score: score,
    status: _comparisonStatus(score, exact: exact),
  );
}

bool departmentMatchesExcludeId(
  DepartmentProfile department,
  String? excludeId,
) {
  if (excludeId == null || excludeId.trim().isEmpty) {
    return false;
  }
  final String needle = excludeId.trim();
  return department.id == needle ||
      department.mutationId == needle ||
      (department.resourceUuid != null && department.resourceUuid == needle) ||
      (department.displayId != null && department.displayId == needle);
}

DepartmentDuplicateCheckResult checkDepartmentDuplicates({
  required String name,
  required String shortName,
  required DepartmentSetupType type,
  required bool isActive,
  required List<DepartmentProfile> existing,
  String? excludeDepartmentId,
  DepartmentProfile? excludeDepartment,
}) {
  final String normalizedName = normalizeDepartmentName(name);
  final String effectiveShortName = resolveDepartmentShortName(name, shortName);
  final String normalizedShortName = normalizeDepartmentName(effectiveShortName);
  var exactNameConflict = false;
  final List<DepartmentSimilarityMatch> matches = <DepartmentSimilarityMatch>[];

  for (final DepartmentProfile department in existing) {
    final bool excluded =
        departmentMatchesExcludeId(department, excludeDepartmentId) ||
        (excludeDepartment != null &&
            (department.id == excludeDepartment.id ||
                department.mutationId == excludeDepartment.mutationId ||
                (excludeDepartment.resourceUuid != null &&
                    department.resourceUuid ==
                        excludeDepartment.resourceUuid) ||
                (excludeDepartment.displayId != null &&
                    department.displayId == excludeDepartment.displayId)));
    if (excluded) {
      continue;
    }

    final String departmentName = normalizeDepartmentName(department.name);
    final String departmentShortName = normalizeDepartmentName(
      resolveDepartmentShortName(department.name, department.shortName),
    );

    final bool nameExact =
        normalizedName.isNotEmpty && departmentName == normalizedName;
    final bool shortNameExact =
        normalizedShortName.isNotEmpty &&
        departmentShortName.isNotEmpty &&
        departmentShortName == normalizedShortName;
    final bool typeExact = department.type == type;
    final bool statusExact = department.isActive == isActive;

    int? nameScore;
    int? shortNameScore;
    int? typeScore;
    int? statusScore;
    final List<String> reasons = <String>[];

    if (normalizedName.isNotEmpty && departmentName.isNotEmpty) {
      nameScore = nameExact
          ? 100
          : nameSimilarityScore(normalizedName, departmentName);
      if (nameExact || nameScore >= departmentSimilarityThreshold) {
        reasons.add('name');
      }
    }

    if (normalizedShortName.isNotEmpty && departmentShortName.isNotEmpty) {
      shortNameScore = shortNameExact
          ? 100
          : nameSimilarityScore(normalizedShortName, departmentShortName);
      if (shortNameExact || shortNameScore >= departmentSimilarityThreshold) {
        reasons.add('short_name');
      }
    }

    typeScore = typeExact ? 100 : 0;
    if (typeExact) {
      reasons.add('department_type');
    }

    statusScore = statusExact ? 100 : 0;
    if (statusExact) {
      reasons.add('status');
    }

    final int score = compositeDepartmentSimilarityScore(
      nameScore: nameScore,
      shortNameScore: shortNameScore,
      typeScore: typeScore,
      statusScore: statusScore,
    );

    final List<DepartmentFieldComparison> fieldComparisons =
        <DepartmentFieldComparison>[
          _fieldComparison(
            field: 'name',
            inputValue: name,
            candidateValue: department.name,
            score: nameScore,
            exact: nameExact,
          ),
          _fieldComparison(
            field: 'short_name',
            inputValue: effectiveShortName,
            candidateValue: resolveDepartmentShortName(
              department.name,
              department.shortName,
            ),
            score: shortNameScore,
            exact: shortNameExact,
          ),
          _fieldComparison(
            field: 'department_type',
            inputValue: type.name,
            candidateValue: department.type.name,
            score: typeScore,
            exact: typeExact,
          ),
          _fieldComparison(
            field: 'status',
            inputValue: isActive ? 'active' : 'inactive',
            candidateValue: department.isActive ? 'active' : 'inactive',
            score: statusScore,
            exact: statusExact,
          ),
          _fieldComparison(
            field: 'display_id',
            candidateValue: department.displayId ?? department.id,
          ),
        ].where(
          (DepartmentFieldComparison entry) =>
              entry.inputValue != null || entry.candidateValue != null,
        )
        .toList(growable: false);

    if (nameExact) {
      exactNameConflict = true;
    }

    final bool isExact = nameExact;
    final bool strongIdentitySignal =
        (nameScore != null && nameScore >= departmentSimilarityThreshold) ||
        (shortNameScore != null &&
            shortNameScore >= departmentSimilarityThreshold);
    final bool compositeSignal = score >= departmentSimilarityThreshold;

    if (!isExact && !strongIdentitySignal && !compositeSignal) {
      continue;
    }

    matches.add(
      DepartmentSimilarityMatch(
        department: department,
        score: score,
        reasons: reasons.isEmpty ? const <String>['name'] : reasons,
        isExact: isExact,
        exactNameConflict: nameExact,
        nameScore: nameScore,
        shortNameScore: shortNameScore,
        typeScore: typeScore,
        statusScore: statusScore,
        fieldComparisons: fieldComparisons,
      ),
    );
  }

  matches.sort(
    (DepartmentSimilarityMatch left, DepartmentSimilarityMatch right) =>
        right.score.compareTo(left.score),
  );

  return DepartmentDuplicateCheckResult(
    exactNameConflict: exactNameConflict,
    similarMatches: matches,
  );
}

File: backend/src/lib/department/department-similarity.js
/**
 * Department duplicate / similarity helpers.
 *
 * Facility-scoped weighted similarity across name, short name, type, and status.
 * Reuses tenant-similarity text scoring primitives.
 */

const {
  SIMILARITY_THRESHOLD,
  normalizeText,
  textSimilarityScore
} = require('@lib/tenant/tenant-similarity');

const NAME_WEIGHT = 50;
const SHORT_NAME_WEIGHT = 20;
const TYPE_WEIGHT = 20;
const STATUS_WEIGHT = 10;

const normalizeDepartmentType = (value) =>
  String(value || '')
    .trim()
    .toUpperCase();

const comparisonStatus = (score, { exact = false } = {}) => {
  if (exact || score === 100) return 'MATCH';
  if (score == null) return 'MISSING';
  if (score >= SIMILARITY_THRESHOLD) return 'SIMILAR';
  return 'DIFFERENT';
};

const buildCandidateSnapshot = (department) => ({
  id: department?.id || null,
  human_friendly_id: department?.human_friendly_id || null,
  display_id:
    department?.display_id || department?.human_friendly_id || department?.id || null,
  tenant_id: department?.tenant_id || null,
  facility_id: department?.facility_id || null,
  name: department?.name || null,
  short_name: department?.short_name || null,
  department_type: department?.department_type || null,
  is_active: department?.is_active !== false
});

const compositeSimilarityScore = (scores = {}) => {
  const weighted = [
    [scores.nameScore, NAME_WEIGHT],
    [scores.shortNameScore, SHORT_NAME_WEIGHT],
    [scores.typeScore, TYPE_WEIGHT],
    [scores.statusScore, STATUS_WEIGHT]
  ];

  let weightedTotal = 0;
  let weightSum = 0;
  for (const [score, weight] of weighted) {
    if (score == null) continue;
    weightedTotal += score * weight;
    weightSum += weight;
  }
  if (!weightSum) return 0;
  return Math.round(weightedTotal / weightSum);
};

const buildFieldComparison = ({
  field,
  inputValue,
  candidateValue,
  score,
  exact = false
}) => ({
  field,
  input_value: inputValue == null || inputValue === '' ? null : String(inputValue),
  candidate_value:
    candidateValue == null || candidateValue === '' ? null : String(candidateValue),
  score: score == null ? null : score,
  status: comparisonStatus(score, { exact })
});

/**
 * @param {Object} params
 * @param {string} params.name
 * @param {string|null|undefined} params.shortName
 * @param {string|null|undefined} params.departmentType
 * @param {boolean|null|undefined} params.isActive
 * @param {Array<Object>} params.existing
 * @param {string|null} [params.excludeDepartmentId]
 */
const checkDepartmentDuplicates = ({
  name,
  shortName,
  departmentType,
  isActive,
  existing = [],
  excludeDepartmentId = null
}) => {
  const normalizedName = normalizeText(name);
  const effectiveShortName =
    String(shortName || '').trim() || String(name || '').trim();
  const normalizedShortName = normalizeText(effectiveShortName);
  const normalizedType = normalizeDepartmentType(departmentType);
  const normalizedActive = isActive === false ? false : true;
  const excludeId = String(excludeDepartmentId || '').trim();

  let exactNameConflict = false;
  const matches = [];

  for (const department of existing) {
    const snapshot = buildCandidateSnapshot(department);
    const departmentId = String(snapshot.id || '').trim();
    const departmentFriendly = String(snapshot.human_friendly_id || '').trim();
    const departmentDisplay = String(snapshot.display_id || '').trim();
    if (
      excludeId &&
      (departmentId === excludeId ||
        departmentFriendly === excludeId ||
        departmentDisplay === excludeId)
    ) {
      continue;
    }

    const departmentName = normalizeText(snapshot.name);
    const departmentShortName = normalizeText(
      snapshot.short_name || snapshot.name
    );
    const departmentTypeValue = normalizeDepartmentType(snapshot.department_type);
    const departmentActive = snapshot.is_active !== false;

    const nameExact = Boolean(normalizedName) && departmentName === normalizedName;
    const shortNameExact =
      Boolean(normalizedShortName) &&
      Boolean(departmentShortName) &&
      departmentShortName === normalizedShortName;
    const typeExact =
      Boolean(normalizedType) &&
      Boolean(departmentTypeValue) &&
      departmentTypeValue === normalizedType;
    const statusExact = departmentActive === normalizedActive;

    let nameScore = null;
    let shortNameScore = null;
    let typeScore = null;
    let statusScore = null;
    const reasons = [];

    if (normalizedName && departmentName) {
      nameScore = nameExact
        ? 100
        : textSimilarityScore(normalizedName, departmentName);
      if (nameExact || nameScore >= SIMILARITY_THRESHOLD) {
        reasons.push('name');
      }
    }

    if (normalizedShortName && departmentShortName) {
      shortNameScore = shortNameExact
        ? 100
        : textSimilarityScore(normalizedShortName, departmentShortName);
      if (shortNameExact || shortNameScore >= SIMILARITY_THRESHOLD) {
        reasons.push('short_name');
      }
    }

    if (normalizedType && departmentTypeValue) {
      typeScore = typeExact ? 100 : 0;
      if (typeExact) {
        reasons.push('department_type');
      }
    }

    statusScore = statusExact ? 100 : 0;
    if (statusExact) {
      reasons.push('status');
    }

    const score = compositeSimilarityScore({
      nameScore,
      shortNameScore,
      typeScore,
      statusScore
    });

    const fieldComparisons = [
      buildFieldComparison({
        field: 'name',
        inputValue: name,
        candidateValue: snapshot.name,
        score: nameScore,
        exact: nameExact
      }),
      buildFieldComparison({
        field: 'short_name',
        inputValue: effectiveShortName,
        candidateValue: snapshot.short_name || snapshot.name,
        score: shortNameScore,
        exact: shortNameExact
      }),
      buildFieldComparison({
        field: 'department_type',
        inputValue: departmentType,
        candidateValue: snapshot.department_type,
        score: typeScore,
        exact: typeExact
      }),
      buildFieldComparison({
        field: 'status',
        inputValue: normalizedActive ? 'active' : 'inactive',
        candidateValue: departmentActive ? 'active' : 'inactive',
        score: statusScore,
        exact: statusExact
      }),
      buildFieldComparison({
        field: 'display_id',
        inputValue: null,
        candidateValue: snapshot.display_id,
        score: null,
        exact: false
      })
    ].filter((entry) => entry.input_value != null || entry.candidate_value != null);

    if (nameExact) {
      exactNameConflict = true;
    }

    const isExact = nameExact;
    const strongIdentitySignal =
      (nameScore != null && nameScore >= SIMILARITY_THRESHOLD) ||
      (shortNameScore != null && shortNameScore >= SIMILARITY_THRESHOLD);
    const compositeSignal = score >= SIMILARITY_THRESHOLD;

    if (!isExact && !strongIdentitySignal && !compositeSignal) {
      continue;
    }

    matches.push({
      ...snapshot,
      score,
      reasons: reasons.length ? reasons : ['name'],
      isExact,
      exactNameConflict: nameExact,
      nameScore,
      shortNameScore,
      typeScore,
      statusScore,
      field_comparisons: fieldComparisons
    });
  }

  matches.sort((left, right) => right.score - left.score);

  return {
    exactNameConflict,
    hasExactConflict: exactNameConflict,
    similarMatches: matches,
    nonExactSimilarMatches: matches.filter((match) => !match.isExact),
    overridableMatches: matches.filter((match) => !match.exactNameConflict)
  };
};

module.exports = {
  SIMILARITY_THRESHOLD,
  NAME_WEIGHT,
  SHORT_NAME_WEIGHT,
  TYPE_WEIGHT,
  STATUS_WEIGHT,
  normalizeDepartmentType,
  compositeSimilarityScore,
  checkDepartmentDuplicates,
  buildCandidateSnapshot
};
